### PR TITLE
create async reader based on vector storage state

### PR DESCRIPTION
**Pull request by generall**

### CODE REVIEW

File: lib/segment/src/vector_storage/tests/async_raw_scorer.rs [hunk 1]
```diff
@@ -79,6 +79,8 @@ fn test_async_raw_scorer(
         test_random_score(&mut rng, &storage, id_tracker.deleted_point_bitslice())?;
     }
 
+    set_async_scorer(false);
```
Comment (timvisee) [inline lib/segment/src/vector_storage/tests/async_raw_scorer.rs:82]: I think this is data racing with concurrent tests.

I'll be able to fix this tomorrow.

**Review by timvisee — CHANGES_REQUESTED** [review]

File: lib/segment/src/vector_storage/memmap_vector_storage.rs [hunk 2]
```diff
@@ -37,12 +38,21 @@ pub fn open_memmap_vector_storage(
...
     distance: Distance,
+) -> OperationResult<Arc<AtomicRefCell<VectorStorageEnum>>> {
+    open_memmap_vector_storage_with_async_io(path, dim, distance, get_async_scorer())
+}
+
+pub fn open_memmap_vector_storage_with_async_io(
+    path: &Path,
+    dim: usize,
+    distance: Distance,
+    with_async_io: bool,
```
Comment (timvisee) [inline lib/segment/src/vector_storage/memmap_vector_storage.rs:49]: What about making `with_async_io` an `Option<bool>` with `.unwrap_or_else(get_async_scorer)`?

Then we can remove the indirect function.

**Review by timvisee — APPROVED** [review]

File: lib/segment/src/vector_storage/tests/async_raw_scorer.rs (hunk 1 above)
Comment (timvisee) [reply lib/segment/src/vector_storage/tests/async_raw_scorer.rs]: Resolved by 8a8469d06d2cab0f0af6bff602414e0847d381f5

File: lib/segment/src/vector_storage/memmap_vector_storage.rs (hunk 2 above)
Comment (generall) [reply lib/segment/src/vector_storage/memmap_vector_storage.rs:49]: in this case would need to refactor all usages. And put None everywhere

**Review by ffuugoo — APPROVED** [review]
LGTM.

**Pull request merged**